pageserver/src/layered_repository/layer_map.rs

```diff
@@ -51,14 +52,37 @@ pub struct LayerMap {
     pub frozen_layers: VecDeque<Arc<InMemoryLayer>>,
 
     /// All the historic layers are kept here
+    historic_layers: RTree<LayerEnvelope>,
 
-    /// TODO: This is a placeholder implementation of a data structure
-    /// to hold information about all the layer files on disk and in
-    /// S3. Currently, it's just a vector and all operations perform a
-    /// linear scan over it.  That obviously becomes slow as the
-    /// number of layers grows. I'm imagining that an R-tree or some
-    /// other 2D data structure would be the long-term solution here.
-    historic_layers: Vec<Arc<dyn Layer>>,
+    /// L0 layers has key range: (Key::MIN..Key::MAX) and locating them using R-Tree search is very inefficient
+    l0_layers: Vec<Arc<dyn Layer>>,
```

Comment:
So, now the `historic_layers` tree keeps only image and l1 layers, right?

We should update all functions that iter over ` historic_layers` to iterate over `l0_layers` too.
I see that `search`, `compact`, and `gc` functions don't do that.

Also, `historic_layers` deserves some comment to not forget it in the future.

Reply:
No, l0 layers are kept both in R-Tree (`historic_layers`) and in vector (`l0_layers`).
Vector is needed to fast locate L0 layers (which can not be done using R-Tree because of full key range.

Reply:
Oh, thank you for explaining.
Please, mention in the `l0_layers` comment that it is just a performance optimization.